websocketOutportDriver/client/wsClient.go

```diff
@@ -0,0 +1,188 @@
+	WSConnClient             WSConnClient
+}
+
+// NewWsClientHandler will create a ws client to receive data from an observer/light client
+func NewWsClientHandler(args *ArgsWsClient) (*client, error) {
+	if args.OperationHandler == nil {
+		return nil, errNilOperationHandler
+	}
+	if args.PayloadParser == nil {
+		return nil, errNilPayloadParser
+	}
+	if args.WSConnClient == nil {
+		return nil, errNilWsConnReceiver
+	}
+	if args.Uint64ByteSliceConverter == nil {
+		return nil, errNilUint64ByteSliceConverter
+	}
+	if len(args.Url) == 0 {
+		return nil, errEmptyUrlProvided
+	}
+
+	urlReceiveData := url.URL{Scheme: "ws", Host: args.Url, Path: data.WSRoute}
+	retryDuration := time.Duration(args.RetryDurationInSec) * time.Second
+
+	return &client{
+		url:                      urlReceiveData.String(),
+		blockingOnError:          args.BlockingAckOnError,
+		retryDuration:            retryDuration,
+		wsConn:                   args.WSConnClient,
+		payloadParser:            args.PayloadParser,
+		operationHandler:         args.OperationHandler,
+		uint64ByteSliceConverter: args.Uint64ByteSliceConverter,
+	}, nil
+}
+
+// Start will start the client listening ws process
+func (c *client) Start() {
+	log.Info("connecting to", "url", c.url)
+
+	for {
+		err := c.wsConn.OpenConnection(c.url)
+		if err != nil {
+			log.Warn(fmt.Sprintf("c.openConnection(), retrying in %v...", c.retryDuration), "error", err.Error())
+			time.Sleep(c.retryDuration)
+			continue
+		}
+
+		closed := c.listenOnWebSocket()
+		if closed {
+			return
+		}
+	}
+}
+
+func (c *client) listenOnWebSocket() (closed bool) {
+	for {
+		_, message, err := c.wsConn.ReadMessage()
+		if err == nil {
+			c.verifyPayloadAndSendAckIfNeeded(message)
+			continue
+		}
+
+		_, isConnectionClosed := err.(*websocket.CloseError)
+		if !isConnectionClosed {
+			if strings.Contains(err.Error(), closedConnection) {
+				return true
+			}
+			log.Warn("c.listenOnWebSocket()-> connection problem, retrying", "error", err.Error())
+		} else {
+			log.Warn(fmt.Sprintf("websocket terminated by the server side, retrying in %v...", c.retryDuration), "error", err.Error())
+		}
+		return
+	}
+}
+
+func (c *client) verifyPayloadAndSendAckIfNeeded(payload []byte) {
+	if len(payload) == 0 {
+		log.Error("empty payload")
+		return
+	}
+
+	payloadData, err := c.payloadParser.ExtractPayloadData(payload)
+	if err != nil {
+		log.Error("error while extracting payload data: " + err.Error())
+		return
+	}
+
+	log.Info("processing payload",
+		"counter", payloadData.Counter,
+		"operation type", payloadData.OperationType,
+		"payload", payloadData.Payload,
+		"message length", len(payloadData.Payload),
+	)
+
+	function, ok := c.operationHandler.GetOperationHandler(payloadData.OperationType)
+	if !ok {
+		log.Warn("invalid operation", "operation type", payloadData.OperationType.String())
+		c.sendAckIfNeeded(payloadData, true)
+		return
+	}
+
+	err = function(payloadData.Payload)
```

Comment:
this can be refactored by sending the error as a parameter in the sendAckIfNeeded function. 
You will not need the bool flag and the code will be smaller.
```go
err := c.processor.ProcessPayload(payloadData)
c.sendAckIfNeeded(payloadData, err)
```

the print for the error can be handled in the sendAckIfNeeded function as it will know about the error, anyway